Proc macros: create types, attributes, and ops

urso opened the pull request:
This PR introduces the `def_type`, `def_attribute` and `def_op` proc macros to create IR entities.

The proc-macro code lives in the `crates/derive` folder. I updated the `Cargo.toml` to become a workspace. Any new crates in the `crates` folder will be part of the workspace automatically (I'm thinking to work on a cranelift crate).

The `ci` job has been updated to use `cargo <...> --workspace` to ensure that the checks are also run on the derive crate.

The `impl_x` macros have been removed from the code and all it's uses have been replaced by the new proc-macros to ensure that we have only one way to create type objects + this verifies that the macros work as expected. We do have some tests in the `derive` crate, but the tests only verify the token stream to match the expected expanded macro contents. We still rely on the tests in the pliron crate to ensure everything works as expected.

All macros follow the same pattern:
1. Each derive proc-macro lives in its own module named `derive_x`.
2. parse the input into an `XInput` type. The parser might run some checks on the input already. Additional attributes that allow users to configure the outcome are parsed into the module local `Attr` type.
4. derive a typed representation of the constructs we want to generate from the input. Like `struct DefOp`. We use `From` or `TryFrom` traits to drive the conversion. The later would run some additional checks on the input or produced output to ensure that we can generate proper code (the IR printer derive macro to be added in a follow up uses that approach)
5. convert the typed representation into a TokenStream

For example `def_op` is defined as:

```
pub(crate) fn def_op(input: impl Into<TokenStream>) -> Result<TokenStream> {
    let input = syn::parse2::<DefOpInput>(input.into())?;
    let p = DefOp::from(input);
    Ok(p.into_token_stream())
}
```

with

```
struct DefOpInput {
    attrs: Attrs,
    ident: syn::Ident,
}

struct DefOp {
    input: DefOpInput,
    verifiers: VerifiersRegister,
}
```

The `VerifiersRegiser` is another shared "construct" for creating the `inventory` struct and implementation. 

I hope that approach is extensible enough such that we can add derive for Verify, Parse, Printers or even more attributes for builders and marker OpInterfaces/AttributeInterfaces.

Sorry for the big PR. The 3 macros reimplement exactly the same original macros. The code itself is quite 'mechanical' and quite similar between the 3 macros.

vaivaswatha commented:
>The proc-macro code lives in the crates/derive folder. I updated the Cargo.toml to become a workspace. Any new crates in the crates folder will be part of the workspace automatically (I'm thinking to work on a cranelift crate)

My idea was *this* repo (pliron) only hold code for the core infrastructure, and others can just use it as a `crates.io` dependence. In fact, I want even the `llvm` dialect to eventually be a separate repo. This repo should just contain the IR infrastructure and the `builtin` dialect. So if you're working on a cranelift crate, you'll just need to add pliron as a dependence in your Cargo.toml and that's it. You don't need to worry about creating PRs here. I believe this provides more freedom and flexibility to users of `pliron`. What do you think?

WIth that in mind, do we need a separate crate for this derive? given that it's an integral part of the IR infrastructure, and if we do, will the `crates/...` directory structure make sense?

vaivaswatha commented on crates/derive/src/attr.rs:
```diff
@@ -0,0 +1,149 @@
+use syn::{Expr, ExprLit, Lit, Result};
+
+// Common trait all attributes should implement.
+pub(crate) trait Attribute
```
Just to confirm, `Attribute` here can be any of `pliron`'s `Op`, `Attribute` or `Type` right?

vaivaswatha commented on src/attribute.rs:
````diff
@@ -252,101 +253,16 @@ pub fn attr_parser<'a>(
 ///
 /// Usage:
 /// ```
+/// #[def_attribute]
````
Can we not do `#[def_attribute = "dialect.name"]`? That would be so much simpler. I saw in your implementation (just a quick glance) that this can be done, so I'm just confirming.

vaivaswatha replied:
> Can we not do `#[def_attribute = "dialect.name"]`? That would be so much simpler. I saw in your implementation (just a quick glance) that this can be done, so I'm just confirming.

I just tried this out, and this doesn't work. Would it be possible to have this instead of two separate macros?

urso replied:
Not sure. Maybe something like `#[def_type("dialect.name")]` should work. Let me give it a try. In case it works I will remove the `dialect` and `type_name` attributes.

vaivaswatha replied:
> Not sure. Maybe something like `#[def_type("dialect.name")]` should work. Let me give it a try. In case it works I will remove the `dialect` and `type_name` attributes.

Thank you.

vaivaswatha commented on crates/derive/src/lib.rs:
````diff
@@ -0,0 +1,100 @@
+///
+/// Usage:
+///
+/// ```ignore
````
Is there any way to avoid having the user declare an empty `struct` ? In my design earlier, I had the user just provide the name. 
1. Having the user provide an empty struct, but later add an `op` field is kind of confusing for the user.
2. Are there checks to ensure that the user doesn't add extra fields? Ideally `Op` should just have that single field containing the `Operation` and nothing else. They're supposed to be lightweight wrappers around an `Operation`.

urso replied:
Yeah, that part is not ideal. One could update the proc macro to accept a type-name and then create the struct for them. Similar to the original macro.

On the other hand I kinda like haveing the proc macro look like an annotation + having a `pub struct MyOp;`.

Plus IMO the doc strings like more natural using that style.

> Are there checks to ensure that the user doesn't add extra fields? Ideally Op should just have that single field containing the Operation and nothing else. They're supposed to be lightweight wrappers around an Operation.

Currently not, but it is easy to add checks.

vaivaswatha replied:
>On the other hand I kinda like haveing the proc macro look like an annotation + having a `pub struct MyOp;`.

Can we restrict the second part to *always* be `pub struct MyOp;` (exactly that, and not `pub struct MyOp {}` or `pub struct MyOp()` or anything else). In that case, I'm open to this syntax. I don't know if that's easy for you to implement, so if it isn't easy, let me know, we can discuss what to do.

>Plus IMO the doc strings like more natural using that style.

I agree.

vaivaswatha reviewed (commented):
A few quick comments after a quick glance. I'll also try out your branch later today.

Thank you @urso .

vaivaswatha commented:
btw, does having proc macros help you with the printer and parser derives? (compared to the existing declarative macros)?

vaivaswatha commented:
> btw, does having proc macros help you with the printer and parser derives? (compared to the existing declarative macros)?

As a user, using the existing declarative macro seems simpler to me right now. 
1. I don't need to define a `struct` explicitly (thus discouraging use of any fields inside).
2. Simpler implementations. I don't mind the added complexity of the procedural macros if they'll help you with the printer and parser derives. But if those derives can be done with just the existing declarative macros, then I don't see the need for these procedural ones.
3. It would be really nice if we could have a single macro instead of two to define each op/type/attr.

urso commented:
> My idea was for this repo (pliron) to only hold code for the core infrastructure, and others can just use it as a crates.io dependence. In fact, I want even the llvm dialect to eventually be a separate repo. This repo should just contain the IR infrastructure and the builtin dialect. So if you're working on a cranelift crate, you'll just need to add pliron as a dependence in your Cargo.toml and that's it. You don't need to worry about creating PRs here. I believe this provides more freedom and flexibility to users of pliron. What do you think?

Sounds good to me. Due to the `llvm` part in the repo I have had the impression that you wanted to collect dialects right in the repo.

> With that in mind, and that the derive marcros you're introducing form an integral part of the IR infrastructure, do we need a separate crate for it? and if we do, will the crates/... directory structure still make sense?

Yes, we need a separate crate. This is because proc macros live in their own crate. But if we don't plan to have more crates I think it makes sense to follow a similar approach to the [`serde`](https://github.com/serde-rs/serde) repository. `serde` uses a workspace but only contains the core and the derive crates. The `serde` crate optionally reexports the `serde_derive` crate.

As we don't need multiple crates I'm thinking to rename `crate/derive` to `pliron_derive`. WDYT?

urso commented:
> btw, does having proc macros help you with the printer and parser derives? (compared to the existing declarative macros)?

Yes. Originally I didn't have these proc macros. But then I ran into the issue the Type/Attributes and Ops are somewhat different in how they store the fields. Types/attributes are more "open" and use the Rust struct, whereas Ops must be extended via Attributes and otherwise are relative strict in what they support as they only given meaning to an `Operation`.
In order to deal with the differences one either has to introduce attributes to pass additional configurations to the derive Macros or alternative we have to create multiple derive macros like `PrintableOp`, `PrintableType`, and `PrintableAttribute`.

As the `if_format` is already an attributed I opted to introduce the hidden attribute `[ir_kind="..."]`. The `[def_x]` macros not only generate the implementation for the struct to come, but they also inject the `[ir_kind="op"]` attribute into the tokenstream to inform the `Printable` derive macro what kind of IR object we are dealing with.

Using it for a while I kinda like the proc macro approach because it gives us a way to introduce more options to introduce builders (e.g. `new_unlinked`, ....) or generate getters and setters for attributes.

> As a user, using the existing declarative macro seems simpler to me right now.

Well, this is my personal opinion, so take it with a grain of salt.

As a consumer of a library I presonaly prefer the annotation-like style that proc macros give me. Using an `impl_x` macro also feels a little weird to me.

> 1. I don't need to define a struct explicitly (thus discouraging use of any fields inside).

Personally I prefer to have a `pub struct MyOp` instead of some macro generating the `pub struct` for me. This way it is explicitly in the code that a type is declared. Plus (at least I do this sometimes) a `grep "struct X"` always works.
I also prefer the doc string to be on the struct directly instead of embedding it in the macro call (not sure if that was for a technical reason).

I agree it is somewhat odd that we require an empty struct for `Op`, or alternatively a struct with `Ptr<Operation>` field.

In the end I don't find any of the 2 macro types very nice. It is kind of a compromise solution.

> 2. Simpler implementations. I don't mind the added complexity of the procedural macros if they'll help you with the printer and parser derives. But if those derives can be done with just the existing declarative macros, then I don't see the need for these procedural ones.

That's true. Unfortunately the proc macros add some complexity (although I don't find declarative macros very readable either 😅). I also wonder if we extend the proc macro based functionality to some extend by adding more attributes (more options similar to TableGen?).

As I said already. I originally introduced these macros to help me pass additional information to the derive macro. Alternatively we could develop 3 derive macros for printers and parsers each, which I found a little less convenient to use. Plus I would also like to introduce derive macros for `Verify` in the future. So the idea of `[def_x]` is also to keep the total number of macros to introduce in check.

> 3. It would be really nice if we could have a single macro instead of two to define each op/type/attr.

Agreed. Let me see if I can get something like `[def_op = "..."]` or `[def_op("...")]` working.

vaivaswatha commented:
> > btw, does having proc macros help you with the printer and parser derives? (compared to the existing declarative macros)?
> 
> Yes. Originally I didn't have these proc macros. But then I ran into the issue the Type/Attributes and Ops are somewhat different in how they store the fields. Types/attributes are more "open" and use the Rust struct, whereas Ops must be extended via Attributes and otherwise are relative strict in what they support as they only given meaning to an `Operation`. In order to deal with the differences one either has to introduce attributes to pass additional configurations to the derive Macros or alternative we have to create multiple derive macros like `PrintableOp`, `PrintableType`, and `PrintableAttribute`.
> 
> As the `if_format` is already an attributed I opted to introduce the hidden attribute `[ir_kind="..."]`. The `[def_x]` macros not only generate the implementation for the struct to come, but they also inject the `[ir_kind="op"]` attribute into the tokenstream to inform the `Printable` derive macro what kind of IR object we are dealing with.
> 
> Using it for a while I kinda like the proc macro approach because it gives us a way to introduce more options to introduce builders (e.g. `new_unlinked`, ....) or generate getters and setters for attributes.
> 
> > As a user, using the existing declarative macro seems simpler to me right now.
> 
> Well, this is my personal opinion, so take it with a grain of salt.
> 
> As a consumer of a library I presonaly prefer the annotation-like style that proc macros give me. Using an `impl_x` macro also feels a little weird to me.
> 
> > 1. I don't need to define a struct explicitly (thus discouraging use of any fields inside).
> 
> Personally I prefer to have a `pub struct MyOp` instead of some macro generating the `pub struct` for me. This way it is explicitly in the code that a type is declared. Plus (at least I do this sometimes) a `grep "struct X"` always works. I also prefer the doc string to be on the struct directly instead of embedding it in the macro call (not sure if that was for a technical reason).
> 
> I agree it is somewhat odd that we require an empty struct for `Op`, or alternatively a struct with `Ptr<Operation>` field.
> 
> In the end I don't find any of the 2 macro types very nice. It is kind of a compromise solution.
> 
> > 2. Simpler implementations. I don't mind the added complexity of the procedural macros if they'll help you with the printer and parser derives. But if those derives can be done with just the existing declarative macros, then I don't see the need for these procedural ones.
> 
> That's true. Unfortunately the proc macros add some complexity (although I don't find declarative macros very readable either 😅). I also wonder if we extend the proc macro based functionality to some extend by adding more attributes (more options similar to TableGen?).
> 
> As I said already. I originally introduced these macros to help me pass additional information to the derive macro. Alternatively we could develop 3 derive macros for printers and parsers each, which I found a little less convenient to use. Plus I would also like to introduce derive macros for `Verify` in the future. So the idea of `[def_x]` is also to keep the total number of macros to introduce in check.
> 
> > 3. It would be really nice if we could have a single macro instead of two to define each op/type/attr.
> 
> Agreed. Let me see if I can get something like `[def_op = "..."]` or `[def_op("...")]` working.

Thank you @urso. I see your point, so let's go with the procedural macros. I only hope it's possible to get the `[def_op("...")]` or `[def_op="..."]` syntax working, otherwise it's ugly.

>Personally I prefer to have a pub struct MyOp instead of some macro generating the pub struct for me.

Although I prefer that it's generated, I don't mind it if what can be declared by the user is controlled, like exactly `pub struct MyOp;` and nothing else. Is that doable? (my knowledge on macros is poor, so it's hard for me to judge what's easy and what's not).

vaivaswatha commented:
> > My idea was for this repo (pliron) to only hold code for the core infrastructure, and others can just use it as a crates.io dependence. In fact, I want even the llvm dialect to eventually be a separate repo. This repo should just contain the IR infrastructure and the builtin dialect. So if you're working on a cranelift crate, you'll just need to add pliron as a dependence in your Cargo.toml and that's it. You don't need to worry about creating PRs here. I believe this provides more freedom and flexibility to users of pliron. What do you think?
> 
> Sounds good to me. Due to the `llvm` part in the repo I have had the impression that you wanted to collect dialects right in the repo.
> 
> > With that in mind, and that the derive marcros you're introducing form an integral part of the IR infrastructure, do we need a separate crate for it? and if we do, will the crates/... directory structure still make sense?
> 
> Yes, we need a separate crate. This is because proc macros live in their own crate. But if we don't plan to have more crates I think it makes sense to follow a similar approach to the [`serde`](https://github.com/serde-rs/serde) repository. `serde` uses a workspace but only contains the core and the derive crates. The `serde` crate optionally reexports the `serde_derive` crate.
> 
> As we don't need multiple crates I'm thinking to rename `crate/derive` to `pliron_derive`. WDYT?

Sounds good to me.

urso commented:
Quick updates:

1. I managed to implement the `[def_op("dialect.op_name")]` style. This also simplified quite some code and the PR is like 400 lines less now. \o/

2. The `def_op` macro will now fail if the user did create a struct with fields.

3. moved the `derive` macro from `crate/derive` to `pliron-derive`.

4. Updated macro tests to use `expect_test`.

> Although I prefer that it's generated, I don't mind it, if what can be declared by the user is controlled, like exactly pub struct MyOp; and nothing else. Is that doable? (my knowledge on macros is poor, so it's hard for me to judge what's easy and what's not).

The `syn::DeriveInput` type carries all attributes (including other derive macros), visibility (pub), data type and per field information (type name, per field attributes) and so on. I'm not to keen to do to many modifications on the data structure, but for compile time checks we have everything available.
All `def_x` macros will check that the user using using a struct + the `def_op` macro will also check that the struct is empty.

Do you think it makes sense to also check that the user has not declared any generics on the type?

For example something like this might be hard to interpret in the context of Pliron, no?

```
#[def_attribute('dialect.generic')]
struct MyAttribute<T>(T>;
```

vaivaswatha commented:
> Quick updates:
> 
> 1. I managed to implement the `[def_op("dialect.op_name")]` style. This also simplified quite some code and the PR is like 400 lines less now. \o/
> 2. The `def_op` macro will now fail if the user did create a struct with fields.
> 3. moved the `derive` macro from `crate/derive` to `pliron-derive`.
> 4. Updated macro tests to use `expect_test`.

Awesome ! I haven't looked at the code itself yet, but will do so soon. I'll also respond to your discussion thread and Issue later today. Thank you again @urso . 

 
> > Although I prefer that it's generated, I don't mind it, if what can be declared by the user is controlled, like exactly pub struct MyOp; and nothing else. Is that doable? (my knowledge on macros is poor, so it's hard for me to judge what's easy and what's not).
> 
> The `syn::DeriveInput` type carries all attributes (including other derive macros), visibility (pub), data type and per field information (type name, per field attributes) and so on. I'm not to keen to do to many modifications on the data structure, but for compile time checks we have everything available. All `def_x` macros will check that the user using using a struct + the `def_op` macro will also check that the struct is empty.


Okay :+1: 

> Do you think it makes sense to also check that the user has not declared any generics on the type?
> 
> For example something like this might be hard to interpret in the context of Pliron, no?
> 
> ```
> #[def_attribute('dialect.generic')]
> struct MyAttribute<T>(T>;
> ```

Yes. I don't see generic attributes, types or ops being supported by the infrastructure in the near future. So it makes sense to add the check against it.
